src/solc_0.8/bundleSandSale/PolygonBundleSandSale.sol

```diff
@@ -0,0 +1,281 @@
+        require(address(_asset) == msg.sender, "only accept asset as sender");
+        require(from == operator, "only self executed transfer allowed");
+        require(ids.length > 0, "need to contains Asset");
+        require(data.length > 0, "data need to contains the sale data");
+
+        (uint256 numPacks, uint256 sandAmountPerPack, uint256 priceUSDPerPack) =
+            abi.decode(data, (uint256, uint256, uint256));
+
+        uint256[] memory amounts = new uint256[](ids.length);
+        for (uint256 i = 0; i < amounts.length; i++) {
+            require(values[i] > 0, "asset transfer with zero values");
+            uint256 amount = values[i] / numPacks;
+            require(amount * numPacks == values[i], "invalid amounts, not divisible by numPacks");
+            amounts[i] = amount;
+        }
+
+        _setupBundle(from, sandAmountPerPack, numPacks, ids, amounts, priceUSDPerPack);
+        return ERC1155_BATCH_RECEIVED;
+    }
+
+    /**
+     * @notice Returns the amount of ETH for a specific amount of USD
+     * @param usdAmount An amount of USD
+     * @return The amount of ETH
+     */
+    function getEtherAmountWithUSD(uint256 usdAmount) public view returns (uint256) {
+        uint256 ethUsdPair = getEthUsdPair();
+        return (usdAmount * 1 ether) / ethUsdPair;
+    }
+
+    /**
+     * @notice Gets the ETHUSD pair from the Medianizer contract
+     * @return The pair as an uint256
+     */
+    function getEthUsdPair() internal view returns (uint256) {
+        bytes32 pair = _medianizer.read();
+        return uint256(pair);
+    }
+
+    function _transferPack(
+        uint256 saleIndex,
+        uint256 numPacks,
+        address to
+    ) internal {
+        uint256 sandAmountPerPack = sales[saleIndex].sandAmount;
+        require(_sand.transferFrom(address(this), to, sandAmountPerPack * numPacks), "Sand Transfer failed");
+        uint256[] memory ids = sales[saleIndex].ids;
+        uint256[] memory amounts = sales[saleIndex].amounts;
+        uint256 numIds = ids.length;
+        for (uint256 i = 0; i < numIds; i++) {
+            amounts[i] = amounts[i] * numPacks;
+        }
+        _asset.safeBatchTransferFrom(address(this), to, ids, amounts, "");
+    }
+
+    function _setupBundle(
+        address from,
+        uint256 sandAmountPerPack,
+        uint256 numPacks,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        uint256 priceUSDPerPack
+    ) internal {
+        require(_sand.transferFrom(from, address(this), sandAmountPerPack * numPacks), "failed to transfer Sand");
+        sales.push(
+            Sale({
+                ids: ids,
+                amounts: amounts,
+                sandAmount: sandAmountPerPack,
+                priceUSD: priceUSDPerPack,
+                numPacksLeft: numPacks
+            })
+        );
+        uint256 saleId = sales.length - 1;
+        emit BundleSale(saleId, ids, amounts, sandAmountPerPack, priceUSDPerPack, numPacks);
```

Comment:
This is a little bit strage, the first returned saleId (zero) can't be used directly in getSalesInfo that expect 1 ?

Reply:
Yes indeed, it's the sale index

Reply:
I can change the functionality so the returned value is right ?

Reply:
I believe it should be like this
```        
uint256 saleId = sales.length;
emit BundleSale(saleId, ids, amounts, sandAmountPerPack, priceUSDPerPack, numPacks);
```
Because BundleSale is waiting for the saleId